jupiterone/data_source_user_group.go

```diff
@@ -0,0 +1,169 @@
+	resp.TypeName = req.ProviderTypeName + "_user_group"
+}
+
+// Schema implements resource.Resource
+func (*userGroupDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "A saved JupiterOne User Group",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"name": schema.StringAttribute{
+				Required:    true,
+				Description: "The name of the user group",
+			},
+			"description": schema.StringAttribute{
+				Computed:    true,
+				Description: "The description of the user group",
+			},
+			"permissions": schema.ListAttribute{
+				Description: "A list of permissions for the user group.",
+				Computed:    true,
+				ElementType: types.StringType,
+			},
+			"query_policy": schema.ListAttribute{
+				Description: "A list of query policy statements for the user group.",
+				Computed:    true,
+				ElementType: types.MapType{
+					ElemType: types.ListType{
+						ElemType: types.StringType,
+					},
+				},
+			},
+		},
+	}
+}
+
+// Read refreshes the Terraform state with the latest data.
+func (d *userGroupDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data UserGroupModel
+
+	// Read Terraform state into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	groups, err := client.GetGroupsByName(ctx, d.qlient, data.Name.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("failed to get user group", err.Error())
+		return
+	}
+
+	// Grab one group that has the same name as the given name
+	if len(groups.IamGetGroupList.Items) == 0 {
+		resp.Diagnostics.AddError("failed to get user group", "no group found with the given name")
+		return
+	}
+
+	var group *client.GetGroupsByNameIamGetGroupListIamGroupPageItemsIamGroup
+
+	for _, groupData := range groups.IamGetGroupList.Items {
+		if groupData.GroupName == data.Name.ValueString() {
+			group = &groupData
+			break
+		}
+	}
+
+	if group == nil {
+		resp.Diagnostics.AddError("failed to get user group", "no group found with the exact given name")
+		return
+	}
+
+	data.Id = types.StringValue(group.Id)
+	data.Description = types.StringValue(group.GroupDescription)
+	data.Permissions = group.GroupAbacPermission.Statement
+
+	// Convert from []map[string]interface{} to []map[string][]string
+	var queryPolicy []map[string][]string
+
+	for _, statementData := range group.GroupQueryPolicy.Statement {
+		var queryPolicyStatement = make(map[string][]string) // Initialize the map
+
+		for key, value := range statementData {
+			// Was unable to parse the []string from the JSON response in any other way.
+			// So we convert the value to a string and then unmarshal it into a []string.
+			stringValue, stringifyError := json.Marshal(value)
```

Comment:
You can do a type assertion to coerce the type maybe (not 100% sure of the context here).  For example:

```
if strVal, ok := value.(string[]); ok{
  // do something with strvalue
}
```

Reply:
This is slick! Saved so many lines

Reply:
Okay, I actually did try this before and it was not working. And I couldn't figure out why so I gave up:

The else block keeps getting hit, and I am getting a kind of `23`, which is a slice.
```
for key, value := range statementData {
	if arrayValue, ok := value.([]string); ok {
		tflog.Trace(ctx, "Query policy has correct value",
			map[string]interface{}{"theVar": value})
		tflog.Trace(ctx, "Query policy has correct value",
			map[string]interface{}{"theKind": reflect.TypeOf(value).Kind()})
		queryPolicyStatement[key] = arrayValue
	} else {
		tflog.Trace(ctx, "Query policy has non-correct value",
			map[string]interface{}{"theVar": value})
		tflog.Trace(ctx, "Query policy has non-correct value",
			map[string]interface{}{"theKind": reflect.TypeOf(value).Kind()})
	}
}
```

When I intentionally create a slice and try to parse it like so. I get a kind of `23`, and it actually works.
```
for key, value := range statementData {
	var value2 interface{} = []string{"foo", "bar", "baz"}
	if arrayValue, ok := value2.([]string); ok {
		tflog.Trace(ctx, "Query policy has correct value",
			map[string]interface{}{"theVar": value})
		tflog.Trace(ctx, "Query policy has correct value",
			map[string]interface{}{"theKind": reflect.TypeOf(value).Kind()})
		queryPolicyStatement[key] = arrayValue
	} else {
		tflog.Trace(ctx, "Query policy has non-correct value",
			map[string]interface{}{"theVar": value})
		tflog.Trace(ctx, "Query policy has non-correct value",
			map[string]interface{}{"theKind": reflect.TypeOf(value).Kind()})
	}
}
```